cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,253 +214,249 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
-      bitmask_type word       = bitmask[word_index(last_bit_index)];
-      bitmask_type slack_mask = set_most_significant_bits(num_slack_bits);
+    // Compute delta due to the following bits in the last word in the range.
+    size_type const last_num_slack_bits = (last_bit_index % word_size_in_bits) == 0
+                                            ? 0
+                                            : word_size_in_bits - intra_word_index(last_bit_index);
+    if (last_num_slack_bits > 0) {
+      bitmask_type const word       = bitmask[word_index(last_bit_index)];
+      bitmask_type const slack_mask = set_most_significant_bits(last_num_slack_bits);
       delta -= __popc(word & slack_mask);
     }
 
+    // Update the null count with the computed delta.
     size_type updated_null_count = *(null_counts + range_id) + delta;
     *(null_counts + range_id)    = updated_null_count;
-
     range_id += blockDim.x * gridDim.x;
   }
 }
 
-// convert [first_bit_index,last_bit_index) to
-// [first_word_index,last_word_index)
-struct to_word_index : public thrust::unary_function<size_type, size_type> {
-  const bool _inclusive                 = false;
-  size_type const* const _d_bit_indices = nullptr;
-
-  /**
-   * @brief Constructor of a functor that converts bit indices to bitmask word
-   * indices.
-   *
-   * @param[in] inclusive Flag that indicates whether bit indices are inclusive
-   * or exclusive.
-   * @param[in] d_bit_indices Pointer to an array of bit indices
-   */
-  __host__ to_word_index(bool inclusive, size_type const* d_bit_indices)
-    : _inclusive(inclusive), _d_bit_indices(d_bit_indices)
-  {
-  }
-
-  __device__ size_type operator()(const size_type& i) const
+/**
+ * @brief Functor that converts bit segment indices to word segment indices.
+ *
+ * Converts [first_bit_index, last_bit_index) to [first_word_index,
+ * last_word_index). The flag `end_of_segment` indicates whether the bit is at
+ * the end of a segment, in which case the word index should be incremented for
+ * bits at the start of a word.
+ */
+struct bit_to_word_index {
+  bit_to_word_index(bool end_of_segment) : end_of_segment(end_of_segment) {}
+  CUDA_DEVICE_CALLABLE size_type operator()(const size_type& bit_index) const
   {
-    auto bit_index = _d_bit_indices[i];
-    return word_index(bit_index) + ((_inclusive || intra_word_index(bit_index) == 0) ? 0 : 1);
+    return word_index(bit_index) + ((!end_of_segment || intra_word_index(bit_index) == 0) ? 0 : 1);
   }
+  bool const end_of_segment;
 };
 
 /**
  * @brief Functor that returns the number of set bits for a specified word
  * of a bitmask array.
  *
  */
-struct word_num_set_bits_functor {
-  word_num_set_bits_functor(bitmask_type const* bitmask_) : bitmask(bitmask_) {}
-  __device__ size_type operator()(size_type i) const
+struct count_set_bits_in_word {
+  count_set_bits_in_word(bitmask_type const* bitmask) : bitmask(bitmask) {}
+  CUDA_DEVICE_CALLABLE size_type operator()(size_type i) const
```

Comment:
For my understanding, could we have gotten away with keeping `__device__` here, since it's a `.cuh` file?

Reply:
I am not entirely sure when we prefer `__device__` and when we prefer `CUDA_DEVICE_CALLABLE`. The only difference in behavior is that `CUDA_DEVICE_CALLABLE` adds `inline`, which seemed like a helpful addition to me.

Reply:
Using  `__device__` here would indeed be _safe_ since this is device code. The switch is good for consistency though, Personally I dislike the implicit inlining and I suspect that its usage is causing inlining of code that developers weren't intentionally requesting, but I'm addressing that in a separate PR modifying these macros and in this case it sounds like you want the inlining so I don't see an issue.